gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpRequestRunnable.java

```diff
@@ -0,0 +1,219 @@
+    this.endpoint = endpoint;
+    this.jsonFactory = jsonFactory;
+    this.headerEnhancers = ImmutableList.copyOf(headerEnhancers);
+    this.callOptions = callOptions;
+    this.request = request;
+    this.methodDescriptor = methodDescriptor;
+    this.httpTransport = httpTransport;
+    this.responseFuture = responseFuture;
+  }
+
+  @Override
+  public void run() {
+    try {
+      try (Writer stringWriter = new StringWriter()) {
+        GenericData tokenRequest = new GenericData();
+
+        HttpRequestFactory requestFactory;
+        GoogleCredentials credentials = (GoogleCredentials) callOptions.getCredentials();
+        if (credentials != null) {
+          requestFactory =
+              httpTransport.createRequestFactory(new HttpCredentialsAdapter(credentials));
+        } else {
+          requestFactory = httpTransport.createRequestFactory();
+        }
+
+        // Create HTTP request body.
+        HttpRequestFormatter requestBuilder = methodDescriptor.getHttpRequestBuilder();
+        methodDescriptor.writeRequestBody(request, stringWriter);
+        stringWriter.close();
+        JsonHttpContent jsonHttpContent = null;
+        if (!Strings.isNullOrEmpty(stringWriter.toString())) {
+          jsonFactory.createJsonParser(stringWriter.toString()).parse(tokenRequest);
+          jsonHttpContent =
+              new JsonHttpContent(jsonFactory, tokenRequest)
+                  .setMediaType((new HttpMediaType("application/json")));
+        }
+
+        // Populate HTTP path and query parameters.
+        Map<String, String> pathParams =
+            requestBuilder.getPathParams(request, methodDescriptor.getPathParams());
+        PathTemplate pathPattern = PathTemplate.create(methodDescriptor.endpointPathTemplate());
+        String relativePath = pathPattern.instantiate(pathParams);
+        GenericUrl url = new GenericUrl(endpoint + relativePath);
+        Map<String, List<String>> queryParams =
+            requestBuilder.getQueryParams(request, methodDescriptor.getQueryParams());
+        for (String queryParam : methodDescriptor.getQueryParams()) {
+          if (queryParams.containsKey(queryParam) && queryParams.get(queryParam) != null) {
+            url.set(queryParam, queryParams.get(queryParam));
+          }
+        }
+
+        HttpRequest httpRequest =
+            requestFactory.buildRequest(methodDescriptor.getHttpMethod(), url, jsonHttpContent);
+        for (HttpJsonHeaderEnhancer enhancer : headerEnhancers) {
+          enhancer.enhance(httpRequest.getHeaders());
+        }
+        httpRequest.setParser(new JsonObjectParser(jsonFactory));
+
+        HttpResponse httpResponse = httpRequest.execute();
+
+        ResponseT response =
+            methodDescriptor.parseResponse(new InputStreamReader(httpResponse.getContent()));
+        responseFuture.set(response);
+
+      } catch (IOException e) {
+        responseFuture.setException(e);
+      }
+    } catch (Exception e) {
```

Comment:
We don't need two levels of catch block here, we can just use one (the outer one)

Reply:
Done.